python-lib/dku_email_client.py

```diff
@@ -0,0 +1,112 @@
+import smtplib
+from email.mime.text import MIMEText
+from email.mime.application import MIMEApplication
+from email.mime.multipart import MIMEMultipart
+
+
+def SmtpConfig():
+    # str, int, bool, bool, str, str
+    return namedtuple("SmtpConfig", "smtp_host, smtp_port, smtp_use_tls, smtp_use_auth, smtp_user, smtp_pass")
+
+
+def AttachmentFile():
+    """ Format for an attachment file with the info broken down nicely for both the client classes """
+    # str, str, str ('application' or 'text'), bytes
+    return namedtuple("AttachmentFile", "file_name, mime_type, mime_subtype, data")
+
+
+class AbstractMessageClient(ABC):
+    def __init__(self, plain_text):
+        self.plain_text = plain_text
+
+
+    def send_email(self, sender, recipient, email_subject, email_body, attachment_files):
+        """
+        :param sender: sender email, str - depending on impl, could be ignored if a sender configured for the channel
+        :param recipient: recipient email, str
+        :param email_subject: str
+        :param email_body: body of either plain text or html, str
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        # Any generic email body decoration goes here
+        email_body_to_send = email_body
+        if self.plain_text:
+            email_body_to_send = email_body + '\n\n'
+
+        self.send_email_impl(sender, recipient, email_subject, email_body_to_send, attachment_files)
+
+    @abstractmethod
+    def send_email_impl(self,  sender, recipient, email_body, email_subject, attachment_files):
+        pass
+
+
+class ChannelClient(AbstractMessageClient):
+    """ Impl using DSS channels that requires DSS 12.6 or later """
+    def __init__(self, plain_text, channel_id):
+        super().__init__(plain_text)
+
+        self.dss_client = dataiku.api_client()
+        self.project_id = dataiku.default_project_key()
+        self.channel = self.dss_client.get_messaging_channel(channel_id)
+
+        logging.info(f"Configured channel messaging client with channel {channel_id} - type: {self.channel.type}, "
+                     f"sender: {self.channel.sender}, plain_text? {self.plain_text}")
+
+    def send_email_impl(self, sender, recipient, email_subject, email_body, attachment_files):
+        """
+        :param sender: sender email, str - is ignored if a sender configured for the channel
+        :param recipient: recipient email, str
+        :param email_subject: str
+        :param email_body: body of either plain text or html, str
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        files = [(a.file_name, a.data, f"{a.mime_type}/{a.mime_subtype}") for a in attachment_files]
+
+        sender_to_use = None if self.channel.sender else sender
+        self.channel.send(self.project_id, [recipient], email_subject, email_body, attachments=files, plain_text=self.plain_text, sender=sender_to_use)
+
+    def quit(self):
```

Comment:
Should be also declared in the abstract class

Reply:
Done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/ee8f64a5f31f04ab7a85974561e6dcee19aa45ad